### ISSUE TITLE
the predict procedure don't support multi-batch

### ISSUE THREAD

**Issue by sharpstill** [zengarden/light_head_rcnn#2]
the predict procedure don't support multi-batch, and we test your code on Nvidia 1080Ti GPU which only achieve 2 images per seconds. Far below than original paper 100 FPS.

**Comment by zengarden** [comment]
Hi, 
About speed:  We test Light-Head with resnet-101 on TITAN XP, it yields 40mAP at 10FPS. Following a common practice, testing time is only counting on network forward~(exclude post-processing).  Maybe u need to modify test.py as follows to see a more reasonable time in testing. BTW, Tensorflow is slow at
[result.txt](https://github.com/zengarden/light_head_rcnn/files/1885983/result.txt)
 starting up, please check the detection speed after enough testing iteration.  My testing log is shown in the attachment.

```
diff --git a/experiments/lizeming/light_head_rcnn.ori_res101.coco.ps_roialign/test.py b/experiments/lizeming/light_head_rcnn.ori_res101.coco.ps_roialign/test.py
index a48e336..7f67a4a 100644
--- a/experiments/lizeming/light_head_rcnn.ori_res101.coco.ps_roialign/test.py
+++ b/experiments/lizeming/light_head_rcnn.ori_res101.coco.ps_roialign/test.py
@@ -12,6 +12,7 @@ from config import cfg, config
 
 import argparse
 import dataset
+import time
 import os.path as osp
 import network_desp
 import tensorflow as tf
@@ -64,10 +65,10 @@ def inference(val_func, inputs, data_dict):
 
     feed_dict = {inputs[0]: resized_img[None, :, :, :], inputs[1]: im_info}
 
-    #st = time.time()
+    st = time.time()
     _, scores, pred_boxes, rois = val_func(feed_dict=feed_dict)
-    #ed = time.time()
-    #print(ed -st)
+    ed = time.time()
+    print(ed -st)
 
     boxes = rois[:, 1:5] / scale
```

About multi-batch testing:  I think it is not much hard to modify test.py to support multi-batch testing. BTW, 

We yield 100FPS with a small backbone network(xception like 145M FLOPs), not ResNet-101 7.6G FLOPs.

**Comment by MaskVulcan** [comment]
@zengarden I have changed the test.py for multi-batch,but I get a problem. How can I distinguish the output of the net to different image.the output of net don't show which image the box belong to? How can I konw it in multi-batch testing?

**Comment by zengarden** [comment]
@MaskVulcan  The outputs should have the same order with inputs.

**Comment by MaskVulcan** [comment]
@zengarden For example, I change the test.py and batch_size is 2, but the output of net ,cls_score, is (1001,81).how can I know the order?each of this belong to which image?

**Comment by MaskVulcan** [comment]
@zengarden oh. I see. Is the rois have the index information?

**Comment by zengarden** [comment]
@MaskVulcan   For multi-batch testing, first you need to set attribute test_batch_per_gpu to number_of_batches(e.g.  2)  in config.py. Then modify test.py to provide feed_dict with multi_image.
If u finished them, scores and rois should have the shape of [2000, 81], [2000, 5].  The first dimension of rois is its gpu_id.

**Comment by MaskVulcan** [comment]
@zengarden I have changed the batch_size to 2, and my feed_dict shape is (2,800,1199,3),(2,6). It seems right but the output is (1001,81).

**Comment by MaskVulcan** [comment]
@zengarden I get this. And  I change the batch size to 2.and I find my output is not very good, The shape of rois is always changing. I print it, just like(2000,5),(577,5)(1844,5).......I am a little confused. I copy the code of get_data_for_singlegpu in your dataset.py to feed multi-batch images to test.

**Comment by zengarden** [comment]
@MaskVulcan  Since there are various number of boxes in each image, the shape of the rois may be changeable . **The first column of rois is its batch_index**,  which is helpful for specifying rois of each image.

**Comment by MaskVulcan** [comment]
@zengarden Yeah. I get it. But I print the image_id. I find that the rois shape is different for the same image when batch_size changed. For example. When single batch, two images have (228,5), (320,5). But in batch size 2, it changed to (230,5)(347,5).And When I change the batch size to 2. The map get 10. It must have some bugs but I cannot find it

**Comment by MaskVulcan** [comment]
@zengarden I have changed the code. I copy the code in your dataset.py for train dataset and use it for test. But I find that when I set batch size to 2, the speed turns to 0.06s/img, map 26. But when I change it to 8, the speed is still 0.06s/img, map turns to 34. It is very strange. Is there any other code I need to change?

**Comment by Mikcal** [comment]
I ran test.py on gtx1070 and got a test speed of 0.18s/img. Is there any problem?

**Comment by lxyyang** [comment]
@MaskVulcan have you finished mutil-batch on prediction? when use mutil-batch, the speed is bigger than one-batch? how many?

**Comment by YoungDav** [comment]
@MaskVulcan @Mikcal  @zengarden @lxyyang 
1. **config.py** test_batch_per_gpu = 2
2. **test.py** input_image.shape [1, H, W, C] -> [2, H, W, C]
but dont work: 
slice index 1 of dimension 0 out of bounds
can you show me more detail about how to change code for mini-batch inference？thanks
